OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ssh/signature/SshSignature.java

```diff
@@ -0,0 +1,148 @@
+ */
+
+package org.sufficientlysecure.keychain.ssh.signature;
+
+import com.android.annotations.Nullable;
+import org.bouncycastle.asn1.ASN1Integer;
+import org.bouncycastle.asn1.ASN1Primitive;
+import org.bouncycastle.asn1.ASN1Sequence;
+import org.bouncycastle.bcpg.PublicKeyAlgorithmTags;
+import org.bouncycastle.util.BigIntegers;
+import org.sufficientlysecure.keychain.pgp.UncachedPublicKey;
+import org.sufficientlysecure.keychain.ssh.key.SshEncodedData;
+import org.sufficientlysecure.keychain.ssh.utils.SshUtils;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.security.NoSuchAlgorithmException;
+
+public class SshSignature {
+
+    private SshEncodedData mSignatureBlob;
+
+    private byte[] mRawSignature;
+
+    private int mAlgorithm;
+
+    private UncachedPublicKey mPublicKey;
+
+    public SshSignature(byte[] rawSignature, int algorithm, @Nullable UncachedPublicKey publicKey) {
+        mSignatureBlob = new SshEncodedData();
+        mRawSignature = rawSignature;
+        mAlgorithm = algorithm;
+        mPublicKey = publicKey;
+    }
+
+    private String getSignatureType() throws NoSuchAlgorithmException {
+        switch (mAlgorithm) {
+            case PublicKeyAlgorithmTags.RSA_SIGN:
+            case PublicKeyAlgorithmTags.RSA_GENERAL:
+                return "ssh-rsa";
+            case PublicKeyAlgorithmTags.ECDSA:
+                return "ecdsa-sha2-"+ SshUtils.getCurveName(mPublicKey);
+            case PublicKeyAlgorithmTags.EDDSA:
+                return "ssh-ed25519";
+            case PublicKeyAlgorithmTags.DSA:
+                return "ssh-dss";
+            default:
+                throw new NoSuchAlgorithmException("Unknown algorithm");
+        }
+    }
+
+    private byte[] getConvertedSignature() throws NoSuchAlgorithmException, IOException {
+        mSignatureBlob.putString(getSignatureType());
+
+        switch (mAlgorithm) {
+            case PublicKeyAlgorithmTags.RSA_SIGN:
+            case PublicKeyAlgorithmTags.RSA_GENERAL:
+                return getFinalSignature(mRawSignature);
+
+            case PublicKeyAlgorithmTags.ECDSA:
+                return getEcDsaSignature();
+
+            case PublicKeyAlgorithmTags.EDDSA:
+                return getFinalSignature(mRawSignature);
+
+            case PublicKeyAlgorithmTags.DSA:
+                return getDsaSignature();
+
+            default:
+                return null;
+        }
+    }
+
+    private byte[] getEcDsaSignature() throws IOException {
+        BigInteger r = getR(mRawSignature);
+        BigInteger s = getS(mRawSignature);
+
+        SshEncodedData rsBlob = new SshEncodedData();
+        rsBlob.putMPInt(r);
+        rsBlob.putMPInt(s);
+
+        return getFinalSignature(rsBlob.getBytes());
+    }
+
+    private BigInteger getR(byte[] rawSignature) throws IOException {
+        ASN1Sequence asn1Sequence = getASN1Sequence(rawSignature);
+        return ASN1Integer.getInstance(asn1Sequence.getObjectAt(0)).getValue();
```

Comment:
do we ever expect those to fail? It's kind of weird for a "convert this to that" method to throw an `IOException`, if this is an "should never happen" deal you can just throw an IllegalStateException